src/consensus/pool/slot_state.rs

```diff
@@ -467,38 +468,52 @@ impl SlotState {
         }
     }
 
-    fn check_safe_to_notar(&mut self, block_hash: &BlockHash) -> SafeToNotarStatus {
+    fn check_safe_to_notar(&mut self, block_hash: BlockHash) -> SafeToNotarStatus {
         // check general voted stake conditions
-        let notar_stake = *self.voted_stakes.notar.get(block_hash).unwrap_or(&0);
+        let notar_stake = *self.voted_stakes.notar.get(&block_hash).unwrap_or(&0);
         let skip_stake = self.voted_stakes.skip;
         if !self.is_weakest_quorum(notar_stake) {
             return SafeToNotarStatus::AwaitingVotes;
         }
         if !self.is_weak_quorum(notar_stake) && !self.is_quorum(notar_stake + skip_stake) {
-            self.pending_safe_to_notar.insert(block_hash.clone());
+            self.pending_safe_to_notar.insert(block_hash);
             return SafeToNotarStatus::AwaitingVotes;
         }
 
         // check parent condition
-        if !self.parents.contains_key(block_hash) {
-            return SafeToNotarStatus::MissingBlock;
-        } else if *self.parents.get(block_hash).unwrap() != ParentStatus::Certified {
-            return SafeToNotarStatus::AwaitingVotes;
+        match self.parents.entry(block_hash.clone()) {
+            Entry::Vacant(_) => return SafeToNotarStatus::MissingBlock,
+            Entry::Occupied(entry) => {
+                if entry.get() != &ParentStatus::Certified {
+                    return SafeToNotarStatus::AwaitingVotes;
+                }
+            }
         }
 
         // check own vote
         let own_id = self.epoch_info.own_id;
         let skip = &self.votes.skip[own_id as usize];
         let notar = &self.votes.notar[own_id as usize];
-        if skip.is_some() || notar.is_some() && &notar.as_ref().unwrap().0 != block_hash {
-            self.sent_safe_to_notar.insert(block_hash.clone());
-            self.pending_safe_to_notar.remove(block_hash);
-            SafeToNotarStatus::SafeToNotar
-        } else {
-            if skip.is_none() && notar.is_none() {
+
+        match (skip, notar) {
```

Comment:
I am not entirely convinced if this patch is better than it was before.  We have more code duplication now but there are fewer subtleties in if statements.  I can go either way.

Reply:
I guess we will just have to leave with it 😄